### ISSUE TITLE
[Tech Debt] Refactor FhirCarePlanGeneratorTest to future proof CI build

### ISSUE THREAD

**Issue by ndegwamartin** [opensrp/fhircore#2402]
**Describe the Issue**

Currently the `FhirCarePlanGeneratorTest` fails due to how HAPI FHIR handles date evaluations using FHIR Path

A work around is required to future proof the unit tests simulating various start periods.

**Resources**
https://build.fhir.org/datatypes.html#Distance

**Comment by ndegwamartin** [comment]
Relevant excerpt from the link above

> **Implementation Note:** If the duration value is specified as a whole number (e.g. 1 month), then when the duration is added or subtracted to a given date(time), the outcome should be rounded to the nearest natural calendar division - e.g. Feb. 1 + 1 mo = March 1, not March 2 or 3 (since 1 month in is defined in UCUM as 30 days).

**Comment by pld** [comment]
This is failing today too, I wrote a conditional that on the first day of the month checks for 6 events and does not add 1 to the index in the final check, but this doesn't feel correct. After re-reading your comment I'm guessing this is less to do with the fact that it is the first of the month and more to do with the June - 4 months being February.

We need to close this issue before we can re-enable this test.

**Comment by pld** [comment]
I'm also suspicious this is actually showing some inconsistency in how tasks are generated depending on the current date, which would be bad.

One way to fix this, which I took a pass at but realized needs a lot more than a hour of thought, if to mock the current time in the system. This feels like the correct approach, and then we could run the test against every day (or some set of edge case days) in the year to be it performs consistently

**Comment by pld** [comment]
Now ignoring the main test causing this failure, let's focus on merging a fix for this first, https://github.com/opensrp/fhircore/pull/2403/files#diff-313ba9ecfbcce23b1a613fc57578626a08a4a6406d91dad43acb5b8fd846efbeR917

**Comment by pld** [comment]
@ndegwamartin to work on implementation plan and communication on nuances of date times to the team, then hand-off for coding

**Comment by ndegwamartin** [comment]
Note from the spec: https://hl7.org/fhir/R4/datatypes.html#Distance

Relevant excerpt from the link above

> **Implementation Note:** If the duration value is specified as a whole number (e.g. 1 month), then when the duration is added or subtracted to a given date(time), the outcome should be rounded to the nearest natural calendar division - e.g. Feb. 1 + 1 mo = March 1, not March 2 or 3 (since 1 month in is defined in UCUM as 30 days).

**Examples of rounding off to closest calendar division as per the spec**

Feb 1 2023 + (30 days) = March 3, 2023 - According to FHIR Path spec it will be 1st March 
Feb 2 2023 + (30 days) = March 2, 2023 - According to FHIR Path spec it will be 2nd March

Feb 1 2024 + (30 days) = March 2, 2024 - According to FHIR Path spec it will be 1st March (leap year)
Feb 2 2024 + (30 days) = March 2, 2024 - According to FHIR Path spec it will be 2nd March

31st Jan, 2023 + (30 days) = March 2, 2023 - According to FHIR Path it will be 28th Feb
31st Jan, 2024 + (30 days) = March 1, 2024 - According to FHIR Path it will be 29th Feb

31st Jan, 2023 + (60 days i.e. '2 month') = April 1, 2023 - According to FHIR Path it will be 31st March
31st Jan, 2024 + (60 days i.e. '2 month') = March 31, 2024 - According to FHIR Path it will be 31st March

**Comment by pld** [comment]
if we can mock the system time, we'd want to have a test with a big loop around it that sets the time to every 2 weeks for a year so we hit the edge cases

**Comment by ndegwamartin** [comment]
yeah we could write tests to cover this in the context of the build however i think additionally it is also a business rules issue and expectations the users, product stakeholders, and the qa team should be aware of and aligned on so as not to report false negatives. it is worth noting that these FHIR Path date calculations are probably used in other places than just the care plan generation and will most likely be used in more places in future.

**Comment by pld** [comment]
Yeah good point and I agree for that side of it, please add a section to the docs

**Comment by f-odhiambo** [comment]
@pld Martin will be picking this up
